polar-core/src/kb.rs

```diff
@@ -757,6 +792,160 @@ impl KnowledgeBase {
+        for (object, shorthand_rules) in &self.resource_blocks.shorthand_rules {
+            for shorthand_rule in shorthand_rules {
+                // We create rule types from shorthand rules in the following scenarios:
+                // 1. when the the third "relation" term points to a related Resource
+                // 2. when the second "implier" term points to a related Resource
+                // 3. when the second "implier" term ponits to a related Actor
+                match shorthand_rule {
+                    ShorthandRule {
+                        head: _,
+                        body: (implier, Some(relation)),
+                    } => {
+                        if let Some(Declaration::Relation(subject)) = self
+                            .resource_blocks
+                            .declarations
+                            .get(object)
+                            .unwrap()
+                            .get(relation)
+                        {
+                            has_relation_rule_types_to_create.insert((
+                                subject.value().as_symbol()?.clone(),
+                                relation.value().as_string()?,
+                                object.value().as_symbol()?.clone(),
+                                true,
+                            ));
+
+                            match self
+                                .resource_blocks
+                                .declarations
+                                .get(subject)
+                                .unwrap()
+                                .get(implier)
+                            {
+                                Some(Declaration::Relation(related_subject)) => {
+                                    has_relation_rule_types_to_create.insert((
+                                        related_subject.value().as_symbol()?.clone(),
+                                        implier.value().as_string()?,
+                                        subject.value().as_symbol()?.clone(),
+                                        true,
+                                    ));
+                                }
+                                Some(Declaration::Role) => {
+                                    has_role_rule_types_to_create.remove(&(
+                                        sym!("Actor"),
+                                        instance!(sym!("String")),
+                                        subject.value().as_symbol()?.clone(),
+                                        false,
+                                    ));
+                                    has_role_rule_types_to_create.insert((
+                                        sym!("Actor"),
+                                        instance!(sym!("String")),
+                                        subject.value().as_symbol()?.clone(),
+                                        true,
+                                    ));
```

Comment:
`HashSet::insert` will overwrite the previous entry — I don't think you need to call `HashSet::remove` first

Reply:
they're identical **except** for the last `required` argument which is being flipped here.

Reply:
Oh... Hash *Set* 😆 

In that case, what about splitting it out into a `HashMap<(Symbol, Instance, Symbol), bool>`?

Reply:
Basically, `remove` & `insert` where everything but a `bool` remains constant feels like a code smell to me